core/src/main/java/org/elasticsearch/cluster/routing/RoutingNode.java

```diff
@@ -37,25 +40,35 @@
 
     private final DiscoveryNode node;
 
-    private final List<ShardRouting> shards;
+    private final Map<ShardId, ShardRouting> shards;
 
     public RoutingNode(String nodeId, DiscoveryNode node) {
-        this(nodeId, node, new ArrayList<ShardRouting>());
+        this(nodeId, node, new LinkedHashMap<>()); // LinkedHashMap to preserve order
     }
 
-    public RoutingNode(String nodeId, DiscoveryNode node, List<ShardRouting> shards) {
+    public RoutingNode(String nodeId, DiscoveryNode node, Map<ShardId, ShardRouting> shards) {
         this.nodeId = nodeId;
         this.node = node;
         this.shards = shards;
     }
 
+    // public for tests
+    public RoutingNode(String nodeId, DiscoveryNode node, Iterable<ShardRouting> shardsCollection) {
+        this(nodeId, node);
+        for (ShardRouting shardRouting : shardsCollection) {
+            ShardRouting previousValue = shards.put(shardRouting.shardId(), shardRouting);
+            assert previousValue == null : "Cannot have two different shards with same shard id on same node";
+        }
+
+    }
```

Comment:
Not sure if I see the utility in keeping this constructor around.  I saw 6 instances where this constructor is used in the tests.  May be better to get rid of this constructor and just create a map instead for those 6 instances, making entries shardId->shardRouting?  Or just have a constructor that takes an `Iterable<ShardRouting>` and automatically creates the internal map out of it in the constructor?